src/line-top-index.js

```diff
@@ -69,53 +69,62 @@ export default class LineTopIndex {
     let oldEnd = traverse(start, oldExtent)
     let newEnd = traverse(start, newExtent)
 
+    let isInsertion = isZero(oldExtent)
     let startNode = this.iterator.insertNode(start)
-    let endNode = this.iterator.insertNode(oldEnd)
+    let endNode = this.iterator.insertNode(oldEnd, !isInsertion)
 
     startNode.priority = -1
     this.bubbleNodeUp(startNode)
     endNode.priority = -2
     this.bubbleNodeUp(endNode)
 
-    if (startNode !== endNode) {
-      let blockIdsToMove = new Set
+    startNode.blockIds.forEach((id) => {
+      if (!this.inclusiveBlockIds.has(id)) return
 
-      startNode.blockIds.forEach((id) => {
-        if (this.inclusiveBlockIds.has(id)) {
-          startNode.blockIds.delete(id)
-          blockIdsToMove.add(id)
-        }
-      })
+      startNode.blockIds.delete(id)
+      startNode.blockHeight -= this.blockHeightsById[id]
+      startNode.distanceFromLeftAncestor.pixels -= this.blockHeightsById[id]
 
-      if (startNode.right) {
-        this.collectBlockIdsForSubtree(startNode.right, blockIdsToMove)
-        startNode.right = null
-      }
+      endNode.blockIds.add(id)
+      endNode.blockHeight += this.blockHeightsById[id]
```

Comment:
Actually, thinking about this more, I think transferring directly in this case is fine. Avoids the overhead of inserting into the set then iterating later when we know exactly what we want to do with them, at the cost of some slight duplication.